doc/hacking/style_guide.md

```diff
@@ -0,0 +1,187 @@
+    function_name(int x, int y) const {
+      ...
+    }
+
+    /**
+     * ...
+     */
+    ...
+
+If the function is a constructor with an initializer list, the `:` goes on the
+same line as the constructor name, and the initializers are listed *one per
+line*, indented by two spaces.
+
+    ClassName::
+    ClassName() :
+      _a("a"),
+      _b("b"),
+      _c("c")
+    {
+    }
+
+Empty function bodies in a .cxx/.I file still contain a single line break (i.e.
+to put `}` on its own line).  In a .h file, an empty function can be given as
+`{}`
+
+Braces
+------
+
+Always put an opening `{` on the same line as the statement that calls for the
+brace.  There is exactly one exception to this: constructors with initializer
+lists.
+
+    SomeClass::
+    SomeClass() :
+      _one(1),
+      _two(2)
+    {
+      if (_one < _two) {
+        ...
+      }
+    }
+
+Comments
+--------
+
+Comments should be properly-formatted English, with proper capitalization and
+punctuation rules.  There are two spaces between sentences.
+
+Note that `//` is preferred over `/*` except in some special circumstances.
+
+Line limits and continuations
+-----------------------------
+
+Lines should be limited to 80 columns.  This is not a hard limit, but if a line
+does go over this length, consider a continuation.
+
+Continuation lines shall be aligned to the same column as appropriate, or
+indented two spaces otherwise:
+
+    if ((a &&
+         b) ||
+	c) {
+      ...
+    }
+
+When choosing where to break a line, these are the preferred places, from best
+to worst:
+
+1. If inside a comment, break the comment to as many lines as appropriate, but
+   only after moving the comment to its own line. (In other words, comments
+   long enough to need 
+2. After a comma in an argument/parameter list.
+3. After a binary operator (leaving the binary operator at the end of the line)
+4. In a function declaration, after the function's return type but before the
+   function's name. (Indented by two spaces.)
+5. In a log statement with `<<`, before a `<<`. (Avoid ending a line with `<<`)
+6. Around a string literal, at a natural word boundary, with a space left on
+   the previous line (e.g. `"This is a "` / `"continued string literal."`)
+7. After the `=` in an assignment operator.
+8. As a last resort, after the opening `(` in a function call.
+
+If none of these rules can apply cleanly, it is okay to have lines extend past
+the 80-column boundary.
+
+Alignment
+---------
+
+TBD, should discuss:
+
+1. Groups of declarations/definitions
+2. Groups of comments
+3. Pointers always being `obj *x;` and never `obj* x;`
+
+Grouping
+--------
+
+Try to group logically-similar lines, separating them with a single blank line.
+
+Modern language features
+------------------------
+
+Panda3D is a C++11 project.  The use of the following modern language features
+is greatly encouraged:
+
+1. `nullptr` over `NULL`
+2. `auto` (when used to store a function return value only)
```

Comment:
Which is preferred: `auto` or adding the typedef? If the latter is preferred, we shouldn't allow `auto` at all since it wouldn't encourage the use of the typedef.

If `auto` is supported by all of our target C++ compilers, might we consider _removing_ the typedefs and using `auto` instead? Or is the reminder in the C++ code about what is being iterated over especially helpful?

Reply:
I think typedef is preferred for member variables.  The typedefs help with clarity in other places as well.  However, they aren't for temporary containers used only inside the function, in which case use of `auto` seems fine to me.

I don't feel strongly about enforcing this strictly, though, and this whole thing becomes moot if we allow range-based for.

Reply:
I think we should make clear where to use auto and where not. My personal opinion:
- Discourage use to declare new variables. For example: `auto x = some_func()`
- Discourage for return types: `auto myfunc() => int {` or with C++14 `auto myfunc() {`
- Encourage to use for range based for loops (once supported): `for (const auto& x : container)`
- Encourage to use for lambdas (once supported): `auto myfunc = []() { return something(); }`
- Encourage to use for iterator types, e.g. `auto begin = mycontainer.begin()` (Although, this should be rarely necessary with range based for loops`

`auto` can make the code much cleaner, but should be used with care

Reply:
I agree with most of that, except in the case of range-based for loops, the member type is generally simple enough to write out and can make the code easier to understand by being explicit.  `auto` could make unclear whether a reference or a local copy of the type is created.

I would also advise that anything that interrogate may need to wrap should not use `auto`, which may be relevant in the future if we want to wrap lambdas for Python code.  (Interrogate will parse `auto`, but it won't deduce the return type automatically except in very simple cases.)